implement add / rename / delete column in data browser [axot: 51]

davidsbailey opened the pull request:
axosoft work item: https://codeorg.axosoft.com/viewitem?id=51&type=features&force_use_number=true

spec: https://docs.google.com/document/d/1UfBgHCisr_8jSt8zHFgJL-XrlLthTrXJQZJwoedGYg4/edit# --> search for "Adding a column"

demo:
![table-column-mutations](https://cloud.githubusercontent.com/assets/8001765/17118717/c8b4395c-5276-11e6-9de1-19781564ff9a.gif)

davidsbailey commented:
Hey @islemaster , would you mind taking a look at some of the latest data browser changes? I'm happy to give you an overview before you dive in if that would be helpful.

islemaster commented on apps/src/applab/dataBrowser/AddColumnButton.jsx:
```diff
@@ -0,0 +1,38 @@
+    );
+  }
+});
+export default Radium(AddColumnButton);
```
Do we expect this add button to change much?  If not, I think there's a more idiomatic way to define it. You're not actually using `Radium` or `redux`, the whole thing is stateless, etc.  Any reason it shouldn't just be expressed this way?

``` jsx
import React from 'react';
import {button as buttonStyle} from './dataStyles';

const containerStyle = {
  float: 'right',
  paddingTop: 10,
  paddingLeft: 10,
  paddingBottom: 10,
  paddingRight: 0
};

export default function AddColumnButton(props) {
  return (
    <div style={containerStyle}>
      <button className="btn" onClick={props.addColumn} style={dataStyles.button}>
        Add column
      </button>
    </div>
  );
}

AddColumnButton.propTypes = {
  addColumn: React.PropTypes.func.isRequired
};
```

davidsbailey replied:
Hi Brad, I agree that this react class is a little light right now. However, the next feature on my plate is to add Import, Export and Delete buttons to this same row, and I had secretly been planning to rename this class to contain the entire top row of buttons (import, export, add column, delete table):
<img width="455" alt="screen shot 2016-07-25 at 4 04 57 pm" src="https://cloud.githubusercontent.com/assets/8001765/17120503/9516733e-5281-11e6-827a-bf38501f2ee6.png">

I've renamed the file and removed some cruft.

islemaster commented on apps/src/applab/dataBrowser/ColumnHeader.jsx:
```diff
@@ -0,0 +1,157 @@
+          {/* TODO(dave): remove 'pull-right' once we upgrade to bootstrap 3.1.0 */}
+          <span className="dropdown pull-right" style={menuStyle}>
+            <a className="dropdown-toggle" data-toggle="dropdown">
+              <i className="fa fa-cog" style={styles.icon} />
```
Nit: Could use [`FontAwesome.jsx`](https://github.com/code-dot-org/code-dot-org/blob/staging/apps/src/templates/FontAwesome.jsx)

islemaster commented on apps/src/applab/dataBrowser/ColumnHeader.jsx:
```diff
@@ -0,0 +1,157 @@
+
+  getInitialState() {
+    return {
+      newName: this.props.columnName,
```
[Props in `getInitialState` is an antipattern.](https://facebook.github.io/react/tips/props-in-getInitialState-as-anti-pattern.html)

davidsbailey replied:
I've initialized `newName` to `''` and modified the render method to use props.columnName if state.newName is not yet set. This seems better than any of the alternatives, but let me know if there is another solution that you prefer.

islemaster commented on apps/src/applab/dataBrowser/ColumnHeader.jsx:
```diff
@@ -0,0 +1,157 @@
+
+  componentDidMount() {
+    if (this.props.isEditing) {
+      ReactDOM.findDOMNode(this.refs.input).select();
```
Because `this.refs.input` is a DOM component and not a composite component, `ReactDOM.findDOMNode` isn't necessary here ([as of React 0.14.0](https://facebook.github.io/react/blog/2015/10/07/react-v0.14.html#dom-node-refs)).

Also, [string refs are considered legacy](https://facebook.github.io/react/docs/more-about-refs.html#the-ref-string-attribute); our styleguide hasn't deprecated them yet, but did you consider callback refs?

islemaster commented on apps/src/applab/dataBrowser/DataTable.jsx:
```diff
@@ -29,6 +41,75 @@ const DataTable = React.createClass({
+  componentWillReceiveProps(nextProps) {
+    // Forget about new columns when switching between tables.
+    if (this.props.tableName !== nextProps.tableName) {
+      this.setState({newColumns: []});
```
Should you also forget about the column being edited when switching between tables?

davidsbailey replied:
great idea -- that lets me change this to this.setState(this.getInitialState()), which generally seems like a good pattern.

islemaster commented on apps/src/applab/dataBrowser/DataTable.jsx:
```diff
@@ -29,6 +41,75 @@ const DataTable = React.createClass({
+  /**
+   * @param {Array|undefined} newColumns
+   */
+  addColumn() {
```
It looks like the documented param here doesn't exist?

islemaster commented on apps/src/applab/dataBrowser/DataTable.jsx:
```diff
@@ -29,6 +41,75 @@ const DataTable = React.createClass({
+        this.props.tableName,
+        oldName,
+        newName,
+        () => {},
```
The empty success callback here makes me wonder if we're missing an intermediate state.  Should the column name field go into some kind of locked/saving state before we call Firebase, and only change to 'not editing' when the save is complete?

davidsbailey replied:
Great point. this needs UX. opened as: https://codeorg.axosoft.com/viewitem?id=84&type=features&force_use_number=true

islemaster commented on apps/src/applab/firebaseStorage.js:
```diff
@@ -342,4 +342,60 @@ FirebaseStorage.populateKeyValue = function (jsonData, overwrite, onSuccess, onE
+ * @param {string} tableName
+ * @param {string} columnName
+ * @param {function()} onSuccess
+ * @param {function(string)} onError
```
Nit: Technically your onError callback might receive an `Error` object, if `JSON.parse` fails.

islemaster commented:
This looks great Dave! Nice clean implementation.

It looks like you're way past this being useful, but it might be good reference: I'm trying to use [Reactabular.js](http://reactabular.js.org/) for the NetSim Log Table.  Even so I'm wrestling with some layout stuff; maybe you can review it when it's ready?

The stateful editing stuff in ColumnHeader might be complex enough to merit a unit test.

Overall LGTM though.  Nice work!

davidsbailey commented:
Thanks for the detailed review, Brad! It's great to learn about some of these best practices. I've addressed all feedback in code unless noted otherwise comments here.

I'd be happy to review the Reactabular change -- I haven't invested too much in HTML tables here, so it'll be a good way to get a sense of whether it'll be useful for me here when I go to implement sorting and such.

davidsbailey commented:
Agreed the column header state needs tests -- noted in https://codeorg.axosoft.com/viewitem?id=59&type=features&force_use_number=true .
